### PR TITLE
Refactor library architecture to remove circular dependancies between libraries

### PR DESCRIPTION
**Pull request by asmodai27**
This is a first step to get the controller to work with platformio 5.0 CLI.
It allows us to build the controller using platformio 5.0 but we still get linker errors for unit tests.

This also includes renaming a variable from a previous peer review that I somehow failed to commit.

### CODE REVIEW

File: software/controller/lib/core/comms.cpp [hunk 1]
```diff
@@ -1,8 +1,7 @@
 #include "comms.h"
 
-#include "algorithm.h"
 #include "hal.h"
-#include "network_protocol.pb.h"
+#include <algorithm>
```
Comment (martukas) [inline software/controller/lib/core/comms.cpp:4]: Remind me, why is the STL algs library now ok while it was not before?

File: software/controller/lib/hal/circular_buffer.h [hunk 2]
```diff
@@ -69,17 +69,17 @@ template <class T, uint N> class CircularBuffer {
   // Returns false if the buffer is full.
   [[nodiscard]] bool Put(T dat) {
     BlockInterrupts block;
-    int h = head_ + 1;
-    if (h > N) {
-      h = 0;
+    int new_head = head_ + 1;
```
Comment (martukas) [inline software/controller/lib/hal/circular_buffer.h:72]: this could even be `auto` to avoid future casting surprises?

**Review by martukas — APPROVED** [review]
This looks pretty straightforward. Just a couple of comments.

File: software/controller/lib/core/comms.cpp (hunk 1 above)
Comment (asmodai27) [reply software/controller/lib/core/comms.cpp:4]: It is not supported by the STM32duino we used at first for the Nucleo board, while it is fully supported by the ST STM32 platform we migrated to when the current HAL became usable.

File: software/controller/lib/hal/circular_buffer.h (hunk 2 above)
Comment (asmodai27) [reply software/controller/lib/hal/circular_buffer.h]: sure, why not?

**Pull request merged**